### DISCUSSION TITLE
Distinguishing between national and local T7 CDA deployments

### DISCUSSION THREAD

**Discussion by adamkorynta** [USACE/cwms-data-api#806] [General Aesthetics]
The CAVI is looking to add a feature that converts JDBC/Oracle extract groups to CDA extract groups. One of the interface options could be to convert to a national or local instance of CDA. On top of that it would be useful to be able to convert a local instance CDA to a national one. This brings up the question of how to determine national vs local.

Some options in order of most to least robust:
- CDA can add a national vs local flag to an endpoint. The authenticated version endpoint seems like a good candidate https://github.com/USACE/cwms-data-api/issues/594
- The client can look at the context and determine `cwms-data` means national, otherwise local
- The client can compare against hard-coded URL strings against CWBI/DMZ URL's

**Comment by MikeNeilson** [comment]
There's a forth option for this that may be more appropriate. The context definitely determines if district or national but if you're looking for can I authenticate the best solution is to look at the Security Schemes defined in the open api definition.

Though I did just realize the current national instance is incorrect as it lists CWMS AAA when it should list none, but that I think is the correct solution for determine the level of access. Per instance the Security Schemes should be defined correctly.

I was planning to correct some of our design decisions about the AccessManager when I get back and can address the actual API output at that time.

**Reply by adamkorynta** [reply]
Maybe I'm not following, are you suggesting that the client pull the openapi and then parse to look for "CWMS AAA" and if it isn't present then it is national?

**Reply by MikeNeilson** [reply]
More or less, though I'm more concerned why the concern is "district vs national" since office is pretty much always required now and the URL is those the distinction.

If the concern is how one can log in and thus write, better to use the API and directly see what's available vs "guessing" as it would be more flexible in the future.

**Reply by adamkorynta** [reply]
This is for automated migration from JDBC sources to CDA, which brought up the idea of migration from district T7 to national. If the user is on a JDBC extract for example, there would be a UI prompt to go against national CDA or an option to pick from a district T7 instance. Then if you have a district T7 instance, a UI prompt could be available to convert to national.

This specific discussion doesn't care about level of access, but seemed like information that could be returned by the request outlined in https://github.com/USACE/cwms-data-api/issues/594

**Reply by MikeNeilson** [reply]
Ah, I see. I don't disagree with #594, and the primary security concern is "dealing with stupid VPD reports once a week at likely a minimum" not actual security.

**Reply by adamkorynta** [reply]
I don't like the idea of parsing the swagger docs over parsing the URL because... well we already see that it wouldn't work correctly. I also don't see any other use case of wanting to parse through swagger docs at the moment at runtime. The "cwms-data" context would be easier as long as that is consistently replaced by "<office>-data" at every non-national instance.

**Reply by MikeNeilson** [reply]
For the context it's always (in regex format):

(district|cwms)-data(-.*)?

Where district is a district or division name and cwms is the literal text "cwms" which indicates national. And the bit at the end is so multiple versions can be used on the same host and is arbitrary text.

If a given system is setup in a way to contradict this design choice it is invalid and we have no practical way to protect against it.


As for the swagger docs not parsing well at runtime... we really need to fix that, but I agree for this particular need not necessary. Would be useful for the setup dialog to present the correct login choices though, One of the design intentions is to support OpenID Connect generically which means there could be more than one openid endpoint, but that's beyond the scope of this discussion.

**Reply by adamkorynta** [reply]
Oh that's a good point about choosing authentication. Swagger does have a java parser library: https://github.com/swagger-api/swagger-parser with an enum for auth types: https://github.com/swagger-api/swagger-core/blob/master/modules/swagger-models/src/main/java/io/swagger/v3/oas/models/security/SecurityScheme.java#L16

**Reply by MikeNeilson** [reply]
One slight caveat, my apikey token schema doesn't match, hence why we need to manually put the apikey word in when using the swagger-ui. Apparently I didn't read the spec fully between swagger/openapi/http when I decided on that name.
